src/goto-programs/goto_coverage.cpp

```diff
@@ -38,64 +62,666 @@ void goto_coveraget::add_false_asserts(goto_functionst &goto_functions)
   goto_programt &goto_program,
-  goto_programt::targett &it)
+  goto_programt::targett &it,
+  const expr2tc &guard)
 {
   goto_programt::targett t = goto_program.insert(it);
   t->type = ASSERT;
-  t->guard = gen_false_expr();
+  t->guard = guard;
   t->location = it->location;
-  t->location.property("assertion");
-  t->location.comment("Instrumentation ASSERT(0) Added");
+  t->location.property("instrumented assertion");
+  t->location.comment(from_expr(ns, "", guard));
   t->location.user_provided(true);
   it = ++t;
-  total_instrument++;
 }
 
 int goto_coveraget::get_total_instrument() const
 {
+  int total_instrument = 0;
+  forall_goto_functions (f_it, goto_functions)
+    if (f_it->second.body_available && f_it->first != "__ESBMC_main")
+    {
+      const goto_programt &goto_program = f_it->second.body;
+      forall_goto_program_instructions (it, goto_program)
+      {
+        if (
+          it->is_assert() &&
+          it->location.property().as_string() == "instrumented assertion" &&
+          it->location.user_provided() == true)
+        {
+          total_instrument++;
+        }
+      }
+    }
   return total_instrument;
 }
 
-// Obtain total assertion instances in goto level via goto-unwind api
+// Count the total assertion instances in goto level via goto-unwind api
 // run the algorithm on the copy of the original goto program
-void goto_coveraget::gen_assert_instance(goto_functionst goto_functions)
+int goto_coveraget::get_total_assert_instance() const
 {
   // 1. execute goto uniwnd
+  int total_assert_instance = 0;
   bounded_loop_unroller unwind_loops;
   unwind_loops.run(goto_functions);
   // 2. calculate the number of assertion instance
+  forall_goto_functions (f_it, goto_functions)
+    if (f_it->second.body_available && f_it->first != "__ESBMC_main")
+    {
+      const goto_programt &goto_program = f_it->second.body;
+      forall_goto_program_instructions (it, goto_program)
+      {
+        if (
+          it->is_assert() &&
+          it->location.property().as_string() == "instrumented assertion" &&
+          it->location.user_provided() == true)
+          total_assert_instance++;
+      }
+    }
+  return total_assert_instance;
+}
+
+std::unordered_set<std::string> goto_coveraget::get_total_cond_assert() const
+{
+  std::unordered_set<std::string> total_cond_assert = {};
+  std::string idf = "";
+
+  forall_goto_functions (f_it, goto_functions)
+  {
+    if (f_it->second.body_available && f_it->first != "__ESBMC_main")
+    {
+      const goto_programt &goto_program = f_it->second.body;
+      forall_goto_program_instructions (it, goto_program)
+      {
+        if (
+          it->is_assert() &&
+          it->location.property().as_string() == "instrumented assertion" &&
+          it->location.user_provided() == true)
+        {
+          idf = it->location.comment().as_string() + "\t" +
+                it->location.as_string();
+          total_cond_assert.insert(idf);
```

Comment:
If this string conceptually is supposed to represent a pair (comment, location), could we make it such? Perhaps a std::pair or an own class that even gives these two fields proper identifiers? Alternatively, could you store `it` directly?

Reply:
Will change to a std::pair